src/ApiException.php

```diff
@@ -138,9 +182,44 @@ public static function createFromRestApiResponse(
             $previous
         );
     }
+    
+    /**
+     * Checks if decoded metadata includes errorInfo message.
+     * @param array $decodedMetadata
+     * @return array $errorInfo {
+     *     @type boolean $containsErrorInfo
+     *     @type string $reason
+     *     @type string $domain
+     *     @type array $errorInfoMetadata
+     * }
+     */
+    public static function containsErrorInfo(array $decodedMetadata)
+    {
+        $errorInfo = [];
+        $errorInfo['containsErrorInfo'] = false;
+        if (empty($decodedMetadata)) {
+            return $errorInfo;
+        }
+        foreach ($decodedMetadata as $value) {
+            $isErrorInfoAny = array_key_exists('@type', $value) && ($value['@type'] === 'google.rpc.errorinfo-bin'
+                || $value['@type'] === 'type.googleapis.com/google.rpc.ErrorInfo');
+            $isErrorInfoArray = array_key_exists('reason', $value) && array_key_exists('domain', $value)
+                && array_key_exists('metadata', $value);
+            if ($isErrorInfoAny || $isErrorInfoArray) {
```

Comment:
I'm a little confused on how this works. I read it as:

      If the error metadata has type ...errorinfo-bin or ...ErrorInfo OR if it has *all* of the keys we like, access all of those keys

By this logic, the type URL here seems unnecessary since an error info message will always have those fields (or most of them, see other comment).

Inversely, when will the `@type` *not` be present or *not* be `errorinfo-bin`/`ErrorInfo`, but still have the fields ?

Reply:
You are totally right - I think I was getting confused between the different types of metadata, but since this function only looks at decoded metadata anyway, then we only need to check for the presence of the keys we want. I also confirmed with Summer that if ErrorInfo exists, then all three keys should be set. I updated the function to just check that `reason` exists for simplicity.